File: java-solution/Mothership.java
import cwinter.codecraft.core.api.*;

class Mothership extends JDroneController {
  static final DroneSpec HARVESTER_SPEC = new DroneSpec().withStorageModules(2);
  static final DroneSpec SOLDIER_SPEC = new DroneSpec().withMissileBatteries(3).withShieldGenerators(1);
  int nHarvesters = 0;


  @Override public void onTick() {
    if (!isConstructing()) {
      if (nHarvesters < 3) {
        buildDrone(new Harvester(this), HARVESTER_SPEC);
        nHarvesters++;
      } else buildDrone(new Soldier(), SOLDIER_SPEC);
    }
  }
}

File: scala-solution/Mothership.scala
import cwinter.codecraft.core.api._


class Mothership extends DroneController {
  var nHarvesters = 0

  override def onTick(): Unit = {
    if (!isConstructing) {
      if (nHarvesters < 3) {
        buildDrone(new Harvester(this), storageModules = 2)
        nHarvesters += 1
      } else buildDrone(new Soldier, missileBatteries = 3, shieldGenerators = 1)
    }
  }
}

